### PR TITLE
Grpc tracing

### PR DESCRIPTION
**Pull request by johnbelamaric**
Enables the tracing to pass through the gRPC client and into the server.

### CODE REVIEW

File: core/dnsserver/server.go [hunk 1]
```diff
@@ -155,6 +155,10 @@ func (s *Server) Address() string { return s.Addr }
 // defined in the request so that the correct zone
 // (configuration and middleware stack) will handle the request.
 func (s *Server) ServeDNS(w dns.ResponseWriter, r *dns.Msg) {
+	s.ServeDNSWithContext(context.Background(), w, r)
+}
+
+func (s *Server) ServeDNSWithContext(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) {
```
Comment (miekg) [inline core/dnsserver/server.go:161]: This needs some docs. Also about where it is used IMO.

Also, I think we should add a TODO that we should find a better solution.

**Comment by johnbelamaric** [comment]
FYI, here's the pretty kind of trace this enables. "dispatcher" and "grpcserver" are two different instances of CoreDNS.

coredns-grpc-trace

File: core/dnsserver/server.go (hunk 1 above)
Comment (johnbelamaric) [reply core/dnsserver/server.go:161]: Ok. We can also wait for 007 on this, it's not a real rush or anything.

Comment (johnbelamaric) [reply core/dnsserver/server.go:161]: We could get a little crazy with this to, to help track down issues in fall through chains like described here https://github.com/coredns/coredns/issues/515. You could bundle the necessary trace data into an EDNS0 local option so you could actually trace through DNS calls not just gRPC.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Grpc tracing ..." ]
>FYI, here's the pretty kind of trace this enables. "dispatcher" and "grpcserver" are two different instances of CoreDNS.
>
>coredns-grpc-trace

Ack. Let's wait for 007, we do one relatively quickly, we don't have some
schedule (yet).

**Comment by johnbelamaric** [comment]
@miekg any more thoughts on this one?

File: core/dnsserver/server.go (hunk 1 above)
Comment (miekg) [reply core/dnsserver/server.go:161]: this all lgtm, with the note that we should go back and fix miekg/dns to make less intrusive

**Pull request merged**